internal/olm/client.go

```diff
@@ -0,0 +1,470 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package olm provides an API to install, uninstall, and check the
+// status of an Operator Lifecycle Manager installation.
+// TODO: move to OLM repository?
+package olm
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	"github.com/operator-framework/operator-sdk/pkg/restmapper"
+
+	olmapiv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	deploymentutil "k8s.io/kubernetes/pkg/controller/deployment/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	olmNamespace = "olm"
+)
+
+var (
+	olmOperatorKey     = types.NamespacedName{Namespace: olmNamespace, Name: "olm-operator"}
+	catalogOperatorKey = types.NamespacedName{Namespace: olmNamespace, Name: "catalog-operator"}
+	packageServerKey   = types.NamespacedName{Namespace: olmNamespace, Name: "packageserver"}
+)
+
+type Client struct {
+	KubeClient client.Client
+	HTTPClient http.Client
+
+	BaseDownloadURL    string
+	BaseReleasesAPIURL string
+}
+
+func ClientForConfig(cfg *rest.Config) (*Client, error) {
+	rm, err := restmapper.NewDynamicRESTMapper(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create dynamic rest mapper: %s", err)
+	}
+
+	sch := scheme.Scheme
+	if err := olmapiv1alpha1.AddToScheme(sch); err != nil {
+		return nil, fmt.Errorf("failed to add OLM types to scheme: %s", err)
+	}
+
+	cl, err := client.New(cfg, client.Options{
+		Scheme: sch,
+		Mapper: rm,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client: %s", err)
+	}
+
+	c := &Client{
+		KubeClient:         cl,
+		HTTPClient:         *http.DefaultClient,
+		BaseDownloadURL:    "https://github.com/operator-framework/operator-lifecycle-manager/releases",
+		BaseReleasesAPIURL: "https://api.github.com/repos/operator-framework/operator-lifecycle-manager/releases",
+	}
+	return c, nil
+}
+
+func (c Client) InstallVersion(ctx context.Context, version string) (*Status, error) {
+	resources, err := c.getResources(ctx, version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resources: %s", err)
+	}
+
+	status := c.getStatus(ctx, resources)
+	if status.HasExistingResources() {
+		return nil, fmt.Errorf("detected existing OLM resources")
```

Comment:
This error should be more verbose, ex. list existing resources, or include "OLM must be completely uninstalled before installation" or something similar.